### PR TITLE
Prevent calls to unimplemented modifiers.

### LINKED ISSUE #11506: ICE on call of unimplemented modifier.
In reference to issue #11468 and its PR #11471 these additionally found ICEs have been isolated.

```solidity
contract A {
    modifier m() virtual { _; }
}
abstract contract B {
    modifier m() virtual;
}
contract C is A, B {
    modifier m() override(A, B) { _; }
    function f() B.m public {}
}
```
Here contract `A` has an implemented modifier `m` and abstract contract `B` has has an unimplemented modifier `m`.
Now, contract `C` inherits from both and overrides `m` with an implemented one.
`C.f()` however explictely references `B.m`, which results into a call to an unimplemented modifier (currently causing ICE in ContractCompiler).

```solidity
abstract contract A {
    modifier m() virtual;
}
contract B is A {
    modifier m() virtual override { _; }
}
contract C is B {
    function f() A.m public {}
}
```
In this case, `A.m` is unimplemented, `B` inheriting from `A` overrides and implements `m`. `C` inheriting from `B` implements a function `f` that statically binds to `A.m`, which also results into an ICE in ContractCompiler.

### PR DESCRIPTION
**Pull request by chriseth**
Fixes https://github.com/ethereum/solidity/issues/11506

### CODE REVIEW

File: libsolidity/analysis/TypeChecker.cpp [hunk 1]
```diff
@@ -639,6 +639,15 @@ void TypeChecker::visitManually(
...
 					);
 			}
+		if (
+			*_modifier.name().annotation().requiredLookup == VirtualLookup::Static &&
+			!modifierDecl->isImplemented()
+		)
+			m_errorReporter.typeError(
+				1835_error,
+				_modifier.location(),
+				"Cannot call unimplemented modifier when referencing it via its contract name."
```
Comment (cameel) [inline libsolidity/analysis/TypeChecker.cpp]: This sounds a bit as if you could call an unimplemented modifier if you referenced it differently. How about this?
```suggestion
				"Cannot call unimplemented modifier. The modifier has no implementation in the referenced contract."
```

**Review by cameel — DISMISSED** [review]
I think it's good if it passes tests. I'd tweak the error message a bit but beyond that everything seems ok.

File: libsolidity/analysis/TypeChecker.cpp (hunk 1 above)
Comment (chriseth) [reply libsolidity/analysis/TypeChecker.cpp]: I wanted to make it clear that you can call the modifier when referencing it without prefixing the contract name.

**Comment by chriseth** [comment]
Rebased.

File: libsolidity/analysis/TypeChecker.cpp (hunk 1 above)
Comment (cameel) [reply libsolidity/analysis/TypeChecker.cpp]: So maybe this?
```suggestion
				"Cannot call unimplemented modifier. The modifier has no implementation in the referenced contract. Refer to it by its unqualified name if you want to call the implementation from the most derived contract."
```

**Review by cameel — DISMISSED** [review]

File: libsolidity/analysis/TypeChecker.cpp (hunk 1 above)
Comment (chriseth) [reply libsolidity/analysis/TypeChecker.cpp]: At first I thought the text is way too long, but probably better to explain more than too little.

**Review by cameel — APPROVED** [review]

File: libsolidity/analysis/TypeChecker.cpp (hunk 1 above)
Comment (cameel) [reply libsolidity/analysis/TypeChecker.cpp]: Maybe we should have something like extra hints for errors. This way the main message could be short and to the point while the user could request a more verbose version with extra explanations.

**Pull request merged**